server/utils.go

```diff
@@ -0,0 +1,183 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/larkox/mattermost-plugin-badges/badgesmodel"
+	"github.com/mattermost/mattermost-server/v5/model"
+)
+
+func areRolesAllowed(userRoles []string, allowedRoles map[string]bool) bool {
+	for ar, b := range allowedRoles {
+		if !b {
+			continue
+		}
+		for _, ur := range userRoles {
+			if ar == ur {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+func canGrantBadge(user model.User, badge badgesmodel.Badge, badgeType badgesmodel.BadgeTypeDefinition) bool {
+	if user.IsSystemAdmin() {
+		return true
+	}
+
+	if badgeType.CreatedBy == user.Id {
+		return true
+	}
+
+	if badge.CreatedBy == user.Id {
+		return true
+	}
+
+	blocked := badgeType.CanGrant.BlockList[user.Id]
+	if blocked {
+		return false
+	}
+
+	if areRolesAllowed(user.GetRoles(), badgeType.CanGrant.Roles) {
+		return true
+	}
+
+	allowed := badgeType.CanGrant.AllowList[user.Id]
+	if allowed {
+		return true
+	}
+
+	return badgeType.CanGrant.Everyone
+}
+
+func canCreateBadge(user model.User, badgeType badgesmodel.BadgeTypeDefinition) bool {
+	if user.IsSystemAdmin() {
+		return true
+	}
+
+	if badgeType.CreatedBy == user.Id {
+		return true
+	}
+
+	blocked := badgeType.CanCreate.BlockList[user.Id]
+	if blocked {
+		return false
+	}
+
+	if areRolesAllowed(user.GetRoles(), badgeType.CanCreate.Roles) {
+		return true
+	}
+
+	allowed := badgeType.CanCreate.AllowList[user.Id]
+	if allowed {
+		return true
+	}
+
+	return badgeType.CanCreate.Everyone
+}
+
+func canEditType(user model.User, badgeType badgesmodel.BadgeTypeDefinition) bool {
+	return user.IsSystemAdmin()
+}
+
+func canEditBadge(user model.User, badge badgesmodel.Badge) bool {
+	return user.IsSystemAdmin() || user.Id == badge.CreatedBy
+}
+
+func canCreateType(user model.User, isPlugin bool) bool {
+	if isPlugin {
+		return true
+	}
+
+	return user.IsSystemAdmin()
+}
+
+func canCreateSubscription(user model.User, channelID string) bool {
```

Comment:
Same here, looks like the `channelID` is not needed at all.

Reply:
Same as before. We may add more permissions, based on, for example, whether the user is a channel admin.

Reply:
Same as before, for me YAGNI applies here too.